subprojects/core/src/main/java/org/gradle/internal/fingerprint/classpath/impl/ClasspathCompareStrategy.java

```diff
@@ -41,56 +43,129 @@ private ClasspathCompareStrategy() {
+            changeState.processChange();
+        }
+        return trackingVisitor.isConsumeMore();
+    }
+
+    private static class TrackingVisitor implements ChangeVisitor {
+        private final ChangeVisitor visitor;
+        private boolean consumeMore = true;
+
+        private TrackingVisitor(ChangeVisitor visitor) {
+            this.visitor = visitor;
+        }
+
+        @Override
+        public boolean visitChange(Change change) {
+            if (consumeMore) {
+                consumeMore = visitor.visitChange(change);
+            }
+            return consumeMore;
+        }
+
+        public boolean isConsumeMore() {
+            return consumeMore;
+        }
+    }
+
+    private static class ChangeState {
+        private Map.Entry<String, FileSystemLocationFingerprint> current;
+        private Map.Entry<String, FileSystemLocationFingerprint> previous;
+        private final boolean includeAdded;
+        private final ChangeVisitor changeConsumer;
+        private final Iterator<Map.Entry<String, FileSystemLocationFingerprint>> currentEntries;
+        private final Map<String, FileSystemLocationFingerprint> currentSnapshots;
+        private final Iterator<Map.Entry<String, FileSystemLocationFingerprint>> previousEntries;
+        private final Map<String, FileSystemLocationFingerprint> previousSnapshots;
+        private final String propertyTitle;
+
+        private ChangeState(String propertyTitle, boolean includeAdded, ChangeVisitor changeConsumer, Map<String, FileSystemLocationFingerprint> currentSnapshots, Map<String, FileSystemLocationFingerprint> previousSnapshots) {
+            this.propertyTitle = propertyTitle;
+            this.includeAdded = includeAdded;
+            this.changeConsumer = changeConsumer;
+            this.currentEntries = currentSnapshots.entrySet().iterator();
+            this.currentSnapshots = currentSnapshots;
+            this.previousEntries = previousSnapshots.entrySet().iterator();
+            this.previousSnapshots = previousSnapshots;
+            this.current = nextEntry(currentEntries);
+            this.previous = nextEntry(previousEntries);
+        }
+
+        void processChange() {
+            if (current == null) {
+                if (previous != null) {
+                    removed();
+                }
+                return;
+            }
+            if (previous == null) {
+                added();
+                return;
+            }
+            FileSystemLocationFingerprint currentFingerprint = current.getValue();
+            FileSystemLocationFingerprint previousFingerprint = previous.getValue();
+            String currentNormalizedPath = currentFingerprint.getNormalizedPath();
+            String previousNormalizedPath = previousFingerprint.getNormalizedPath();
+            if (!currentNormalizedPath.equals(previousNormalizedPath)) {
+                removed();
+                added();
+                return;
+            }
```

Comment:
This is when we have a directory on the classpath, and there are entries in it, right? I don't see any tests covering this case yet.

Reply:
There are already tests in the test class which test this. Basically all the tests which have been there with non-empty relative paths.